### DISCUSSION TITLE
Coherence score of different model represeantion

### DISCUSSION THREAD

**Discussion by TalaN1993** [MaartenGr/BERTopic#2194] [General]
@MaartenGr I  have a question and would like to know your opinion. In a BERTopic model, if we want to compare the coherence of different representations while accepting the other parts as default, is it logical for the coherence of a model like KeyBERT to be higher than that of an OpenAI model? Should we expect the coherence score of OpenAI to always be higher?

**Comment by MaartenGr** [comment]
A couple of things first. OpenAI is not the model itself, so it highly depends on which underlying LLM you are actually choosing. Second, if you are using an LLM, you are generally generating labels and not keywords. Calculating traditional coherence (e.g., "c_v") is therefore not possible since it requires keywords instead. 

That said, if you are generating keywords with an LLM and/or you use a coherence metric that can compare keywords with labels, then you cannot be sure the LLM will always be higher. There are so many factors related to this, such as the data, language, size/quality of the LLM, coherence metric, etc. that you cannot expect these things beforehand.

**Reply by TalaN1993** [reply]
You are right. In fact, I chose two representation models, KeyBERT and GPT-3.5, with the same embedding model, ALL-Mini, and kept the other steps as default.  And I intend to compare the coherence CV of these with each other.

I didn’t quite understand when you said **you generate labels and not keywords**. I would appreciate it if you could explain or guide me to another reference.

**Reply by MaartenGr** [reply]
It might be worthwhile to first research a bit how [coherence scores](https://radimrehurek.com/gensim/models/coherencemodel.html) actually work. More specifically, it is applied on keywords (generated for example through c-TF-IDF or KeyBERT) but not on labels (generated with GPT-3.5). For instance, if you were go use GPT-3.5, you generated labels (e.g., `"Introductions to Large Language Models"`) whereas the others generated keywords (e.g., `["LLM", "model", "nlp", "introduction"]`). 

Coherence score, as linked above, expects keywords and not a single long label.

**Reply by TalaN1993** [reply]
Yes, now I understand what you mean. So, if after receiving the output from GPT, I split it and remove the stop words again, I can convert the labels into keywords. Isn't that right?

**Reply by MaartenGr** [reply]
I really would advise **against** doing that. The labels that are extracted are **not** keywords (even if you split them into words) and not directly extracted from the corpus. Coherence measures the keywords against the corpus, so this really would not be helpful to you. 

Another helpful resource is [this paper](https://svn.aksw.org/papers/2015/WSDM_Topic_Evaluation/public.pdf). It explains how coherence works and allows you to get more intuition about why labels don't work (even if you convert them into separate words).

**Reply by MaartenGr** [reply]
You will also find more information about [coherence here](https://github.com/MaartenGr/BERTopic/issues/90) and my view on this metric.

**Reply by TalaN1993** [reply]
@MaartenGr Hello.I tried to obtain topic diversity and NPMI coherence for my model using OCTIS. However, the challenge I’m facing is related to calculating coherence. I managed to get the values for diversity, but my NPMI output is coming out as ERROR.

ValueError: unable to interpret topic as either a list of tokens or a list of ids

**Reply by MaartenGr** [reply]
@TalaN1993 That seems to be related to OCTIS, so I would advise opening up an issue over there as they know about OCTIS and its internals more than I do. Also, the link I shared previously about my view on the metric also contains some example code I believe.

**Reply by TalaN1993** [reply]
@MaartenGr Hello. I hope you're doing well. I need your guidance. For a part of my research, I need to analyze the output of LDA and topic modeling. After conducting many experiments, I achieved the results I could. But, I have two questions:
1.When I ran the five main steps of BERTopic, I obtained higher diversity and coherence compared to when I included the sixth step. The difference was quite significant. Why does this happen?
2.The diversity of LDA is higher than that of BERTopic, whereas I expected BERTopic to always perform better. Is this possible? And what could be the reason for this?

**Reply by MaartenGr** [reply]
1. It's difficult to say since it's not clear to me which representation step you are using since you referred earlier to two different steps (GTP/KeyBERT). Which is it?
2. That depends, are you comparing an equal number of topics to one another? If so, how did you set the number of topics for BERTopic? For instance, you can't compare two models against one another that have a different number of topics.

**Reply by TalaN1993** [reply]
1)Yes, KeyBERT
2) Yes, I had about 103 topics from the BERTopic ooutput and chose 103 as the number of LDA topic.

**Reply by MaartenGr** [reply]
> 1)Yes, KeyBERT

The KeyBERT methodology is used to improve the representations for human readers but tends to ignore domain-specific words considering it is an embedding-based algorithm. Compared to c-TF-IDF (which might be more difficult to read for humans), it does tend to capture domain-specific entities quite well due to the use of it's bag-of-words.

It is a nice example of why coherence is just not a great evaluation metric by itself. You can add more automatic metrics or look for other techniques, such as human evaluation or perhaps clustering-based metrics. 

> 2) Yes, I had about 103 topics from the BERTopic ooutput and chose 103 as the number of LDA topic.

Without knowing the exact parameters and code, it is difficult to say. The one thing I can say is that one algorithm is seldom to outperform another in all use cases. It might just be the no-free lunch theorem in effect here. 

Also, the diversity metric is not a semantic based method I believe, so there might also be nuances there that are missing. These topic coherence and diversity are not necessarily bad, but we should definitely not consider them a golden standard.

**Reply by TalaN1993** [reply]
Thank you so much for clarification!
